### PR TITLE
fetch: take an s3:// key as written, encode it once on the wire

### PR DESCRIPTION
**Pull request by robobun**
### Problem
- `fetch("s3://bkt/my file.txt")` sends `GET /bkt/my%2520file.txt`. `Bun.file("s3://bkt/my file.txt")` and `S3Client.file("my file.txt")` send `GET /bkt/my%20file.txt`. The same string addresses two different objects. Via `fetch` no spelling reaches a key with a space or a `%`. `dir/../x.txt` collapses to `x.txt` and a tab is dropped.
- `fetch` ran the string through the WHATWG URL serializer (`src/runtime/webcore/fetch.rs:532`), which percent-encodes the key, and the S3 signer (`src/s3_signing/credentials.rs:372`) encoded it again. `new Request("s3://...")` did the same at `src/runtime/webcore/Request.rs:1382`. `Bun.file` never URL-parses the string.

### Fix
- `bun_url::is_s3_url` is one check for an `s3://` string (case-insensitive, the same as `URL::is_s3`, which decides whether `fetch` signs the request). `fetch(string)` and the `Request` constructor skip the URL serializer when it matches and keep the key bytes as written. The signer encodes them once, the same as `Bun.file` and `S3Client.file`.
- Correct because an S3 key is a byte string, not a URL path. `S3Client.file` already works this way, and the docs present `fetch("s3://...")` and `Bun.file("s3://...")` as interchangeable.
- Wire paths before and after, for `s3://bkt/<key>` via `fetch` (`Bun.file` sends the "after" path in both versions):
  `my file.txt`: `/bkt/my%2520file.txt` -> `/bkt/my%20file.txt`
  `ü.txt`: `/bkt/%25C3%25BC.txt` -> `/bkt/%C3%BC.txt`
  `my%20file.txt`: `/bkt/my%2520file.txt` -> `/bkt/my%2520file.txt` (unchanged)
  `dir/../x.txt`: `/bkt/x.txt` -> `/bkt/dir/../x.txt`
  `a\tb`: `/bkt/ab` -> `/bkt/a%09b`
- Verified: `test/js/bun/s3/s3.test.ts` ("s3:// url key encoding", 6 cases, 5 fail on the released bun). Also the rest of `test/js/bun/s3/`, `test/js/web/fetch/fetch.test.ts` and `test/js/web/request/` (no new failures).

### Background
- SigV4 signing builds a canonical request from the percent-encoded key. `encode_uri_component` in `src/s3_signing/credentials.rs` encodes every byte except the RFC 3986 unreserved set, so a key that is already encoded is encoded twice.
- `Bun.file("s3://...")` stores the raw string. `S3::path()` in `src/jsc/webcore_types.rs` runs the non-normalizing `bun_url::URL::parse` on it and takes everything after `s3://`.
- `fetch` needs a normalized href for `http(s)://`. The `s3://` branch never uses the host or path fields. It only reads `url.s3_path()`, the bytes after the scheme.

<details><summary>Notes</summary>

- Migration note: an object written through `fetch(..., { method: "PUT" })` with a key that has a space or non-ASCII byte lives under the doubly encoded key (`report%202026.pdf` for `report 2026.pdf`). After this change `fetch` reads and writes `report 2026.pdf`, the same key `Bun.file` always used. To reach the old object, pass the doubly encoded spelling (`report%25202026.pdf`).
- `fetch(new URL("s3://bkt/a b"))` is unchanged. A `URL` object is already serialized when it reaches `fetch`.
- `fetch("S3://bkt/a b")` is covered too. `Bun.file("S3://...")` is a local path, so the uppercase test row checks `fetch` and `Request` only.
- `new Request("s3://bkt/a b").url` now returns the string as written. `Request.clone()` and `new Request(request)` carry it through.
- Edge cases checked: `s3://`, `s3://bkt` and `s3://bkt/` reject with `ERR_S3_INVALID_PATH` from both `fetch` and `Bun.file`. PUT with a string body, PUT with a `ReadableStream` body and DELETE via `fetch` all send the single-encoded path.
- Self-reviewed: 3 concerns raised, 3 addressed (the `Request` constructor path, one shared check in `bun_url`, a case-insensitive scheme match).
- `test/js/web/request/request-method-getter.test.ts` and `request-clone-leak.test.ts` time out at 5 s in this debug container with and without the change.
</details>

---

**[human-review]** gate passed · iteration 1 · 4 files touched

<details><summary>fails on main (without fix)</summary>

```console
ASAN without fix: 5 failed, 286 skipped
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/pr_gate.xml" "test/js/bun/s3/s3.test.ts"
bun test v1.4.3 (f42e98025)

test/js/bun/s3/s3.test.ts:
failed to connect to the docker API at unix:///var/run/docker.sock; check if the path is correct and if the daemon is running: dial unix /var/run/docker.sock: connect: no such file or directory
(skip) R2 > s3 > fetch > bucket in path > should download file via fetch GET
(skip) R2 > s3 > fetch > bucket in path > should download range
(skip) R2 > s3 > fetch > bucket in path > should check if a key exists or content-length
(skip) R2 > s3 > fetch > bucket in path > should check if a key does not exist
(skip) R2 > s3 > fetch > bucket in path > should be able to set content-type
(skip) R2 > s3 > fetch > bucket in path > should be able to upload large files
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download file via Bun.s3().text()
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range with 0 offset
(skip) R2 > s3 > Bun.S3Client > bucket in path > should check i
... (truncated)

release without fix: 286 skipped
bun test v1.4.3-canary.1 (966da26fe)

test/js/bun/s3/s3.test.ts:
failed to connect to the docker API at unix:///var/run/docker.sock; check if the path is correct and if the daemon is running: dial unix /var/run/docker.sock: connect: no such file or directory
(skip) R2 > s3 > fetch > bucket in path > should download file via fetch GET
(skip) R2 > s3 > fetch > bucket in path > should download range
(skip) R2 > s3 > fetch > bucket in path > should check if a key exists or content-length
(skip) R2 > s3 > fetch > bucket in path > should check if a key does not exist
(skip) R2 > s3 > fetch > bucket in path > should be able to set content-type
(skip) R2 > s3 > fetch > bucket in path > should be able to upload large files
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download file via Bun.s3().text()
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range with 0 offset
(skip) R2 > s3 > Bun.S3Client > bucket in path > should check if a key exists or content-length
(skip) R2 > s3 > Bun.S3Client > bucket in path > should check if a key does not exist
(skip) R2 > s3 > Bun.S3Client > 
... (truncated)
```

</details>

<details><summary>passes on PR (with fix)</summary>

```console
ASAN with fix: 286 skipped
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/pr_gate.xml" "test/js/bun/s3/s3.test.ts"
bun test v1.4.3 (f42e98025)

test/js/bun/s3/s3.test.ts:
failed to connect to the docker API at unix:///var/run/docker.sock; check if the path is correct and if the daemon is running: dial unix /var/run/docker.sock: connect: no such file or directory
(skip) R2 > s3 > fetch > bucket in path > should download file via fetch GET
(skip) R2 > s3 > fetch > bucket in path > should download range
(skip) R2 > s3 > fetch > bucket in path > should check if a key exists or content-length
(skip) R2 > s3 > fetch > bucket in path > should check if a key does not exist
(skip) R2 > s3 > fetch > bucket in path > should be able to set content-type
(skip) R2 > s3 > fetch > bucket in path > should be able to upload large files
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download file via Bun.s3().text()
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range
(skip) R2 > s3 > Bun.S3Client > bucket in path > should download range with 0 offset
(skip) R2 > s3 > Bun.S3Client > bucket in path > should check i
... (truncated)

release with fix: 286 skipped
$ bun scripts/build.ts --profile=release
[configured] bun-profile → bun (stripped) in 1034ms (unchanged)
ninja: Entering directory `/workspace/bun/build/release'
[1/38] gen generated_host_exports.rs
generated_host_exports.rs: 122 exports (host=5, lazy=10, generic=107, rust=0); 242 extern-C blocks audited
[2/38] gen ZigGeneratedClasses.{cpp,h,rs}
Found 2 classes from /workspace/bun/src/jsc/resolve_message.classes.ts
  - ResolveMessage (15 fields)
  - BuildMessage (10 fields)
Found 1 classes from /workspace/bun/src/runtime/api/Archive.classes.ts
  - Archive (4 fields, 1 class fields)
Found 2 classes from /workspace/bun/src/runtime/api/BunObject.classes.ts
  - ResourceUsage (8 fields)
  - Subprocess (20 fields)
Found 1 classes from /workspace/bun/src/runtime/api/cron.classes.ts
  - CronJob (5 fields)
Found 3 classes from /workspace/bun/src/runtime/api/filesystem_router.classes.ts
  - FileSystemRouter (5 fields)
  - FrameworkFileSystemRouter (2 fields)
  - MatchedRoute (8 fields)
Found 1 classes from /workspace/bun/src/runtime/api/Glob.classes.ts
  - Glob (5 fields)
Found 1 classes from /workspace/bun/src/runtime/api/h2.classes.ts
  - H2FrameParser (32 fields)
Found 9
... (truncated)
```

</details>

<details><summary>diff hotspot</summary>

```
src/runtime/webcore/Request.rs | 29 +++++++----------
 src/runtime/webcore/fetch.rs   | 36 +++++++++++----------
 src/url/lib.rs                 |  6 ++++
 test/js/bun/s3/s3.test.ts      | 72 +++++++++++++++++++++++++++++++++++++++---
 4 files changed, 106 insertions(+), 37 deletions(-)
```

</details>

**gate history** · 1 passed · 1 rejected · iteration 1

<details><summary>evidence per changed file</summary>

```
file                            reads  edits  tests
src/runtime/webcore/Request.rs      2      2     24
src/runtime/webcore/fetch.rs        5      4     24
src/url/lib.rs                      3      4     24
test/js/bun/s3/s3.test.ts           1      0     24
```

</details>

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 1:58 AM PT - Sep 6th, 2026</sup></div>

@robobun, your commit 52ab308b814b221627aaea787f1e2a5cdab209dc is building: [`#110959`](https://buildkite.com/bun/bun/builds/110959)

File: test/js/bun/s3/s3.test.ts [hunk 1]
```diff
@@ -2180,3 +2180,57 @@ describe("presigned url signature", () => {
...
+  // fetch("s3://..."), fetch(new Request("s3://...")), Bun.file("s3://...")
+  // and S3Client.file(key) must address the same object: the key bytes after
+  // "s3://bucket/" are taken as written and percent-encoded once on the wire.
+  it.each([
+    ["my file.txt", "/bkt/my%20file.txt"],
+    ["\u00fc.txt", "/bkt/%C3%BC.txt"],
+    ["my%20file.txt", "/bkt/my%2520file.txt"],
+    ["dir/../x.txt", "/bkt/dir/../x.txt"],
+    ["a\tb", "/bkt/a%09b"],
+  ])("fetch, Request, Bun.file and S3Client.file send the same path for key %j", async (key, expectedPath) => {
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/s3/s3.test.ts]: Done in 840ccd1: the block is describe.concurrent now.

File: src/url/lib.rs [hunk 2]
```diff
@@ -189,6 +189,16 @@ pub use whatwg::{
     file_url_from_string, href_from_string, join, origin_from_slice, path_from_file_url,
 };
 
+/// `s3://bucket/key` is a bucket and key bytes, not a WHATWG URL. The key is
+/// taken as written (spaces, non-ASCII, `%XX`, `.`/`..` segments, tabs) and
+/// the S3 signer percent-encodes it once. `fetch`, `Request` and `Bun.file`
+/// all use this check, so the same string addresses the same object. The
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/url/lib.rs]: Done: the doc comment no longer claims Bun.file calls this function (840ccd1, trimmed further in 6a9725e).

File: src/runtime/webcore/fetch.rs [hunk 3]
```diff
@@ -529,22 +529,28 @@ fn fetch_impl<const ALLOW_GET_BODY: bool>(
...
-            return Ok(
-                JSPromise::dangerously_create_rejected_promise_value_without_notifying_vm(
-                    global_this,
-                    err,
-                ),
-            );
-        }
+    //
+    // An `s3://` URL skips the WHATWG serializer: see `bun_url::is_s3_url`.
+    let mut url_proxy_buffer = if bun_url::is_s3_url(&url_str) {
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/fetch.rs:532]: Agreed. is_s3_url now matches S3:// as well, the same rule as URL::is_s3, and the test has an S3:// row (840ccd1).

File: src/runtime/webcore/Request.rs [hunk 4]
```diff
@@ -1379,24 +1379,29 @@ impl Request {
...
-            // so we must increment the reference count before calling it.
-            let err = global_this.err_invalid_url(format_args!(
-                "Failed to construct 'Request': Invalid URL \"{}\"",
-                req.url.get()
-            ));
-            bail!(Err(global_this.throw_value(err)));
-        }
+        // An `s3://` URL keeps its key bytes as written, so that
+        // `fetch(new Request(url))` addresses the same object as `fetch(url)`
+        // and `Bun.file(url)`: see `bun_url::is_s3_url`.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/Request.rs]: Trimmed to one line in 6a9725e.

File: src/runtime/webcore/Request.rs [hunk 5]
```diff
@@ -1379,24 +1379,29 @@ impl Request {
...
+        if !bun_url::is_s3_url(req.url.get()) {
+            let href = bun_url::href_from_string(req.url.get());
+            if href.is_empty() {
+                // globalThis.throw can cause GC, which could cause the above string to be freed.
+                // so we must increment the reference count before calling it.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/Request.rs]: Trimmed to one line in 6a9725e.

File: src/runtime/webcore/Request.rs [hunk 6]
```diff
@@ -1379,24 +1379,29 @@ impl Request {
...
-        // hrefFromString increments the reference count if they end up being
-        // the same
-        //
-        // we increment the reference count on usage above, so we must
-        // decrement it to be perfectly balanced.
+            // hrefFromString increments the reference count if they end up being
+            // the same
+            //
+            // we increment the reference count on usage above, so we must
+            // decrement it to be perfectly balanced.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/Request.rs]: Trimmed to one line in 6a9725e.

File: src/runtime/webcore/fetch.rs [hunk 7]
```diff
@@ -529,22 +529,28 @@ fn fetch_impl<const ALLOW_GET_BODY: bool>(
...
-            );
-            return Ok(
-                JSPromise::dangerously_create_rejected_promise_value_without_notifying_vm(
-                    global_this,
-                    err,
-                ),
-            );
-        }
+    //
+    // An `s3://` URL skips the WHATWG serializer: see `bun_url::is_s3_url`.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/fetch.rs]: Trimmed to one line in 6a9725e.

File: src/url/lib.rs [hunk 8]
```diff
@@ -189,6 +189,19 @@ pub use whatwg::{
...
 };
 
+/// `s3://bucket/key` is a bucket and key bytes, not a WHATWG URL. The key is
+/// taken as written (spaces, non-ASCII, `%XX`, `.`/`..` segments, tabs) and
+/// the S3 signer percent-encodes it once. `fetch` and `Request` skip the
+/// WHATWG serializer when this is true, so the string reaches the signer
+/// with the same key bytes as `Bun.file("s3://...")`, which checks the
+/// `s3://` prefix of its path bytes. The scheme match is case-insensitive,
+/// the same as [`URL::is_s3`], which decides whether `fetch` signs the
+/// request.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/url/lib.rs]: Trimmed to one line in 6a9725e.

File: src/url/lib.rs [hunk 9]
```diff
@@ -189,6 +189,14 @@ pub use whatwg::{
     file_url_from_string, href_from_string, join, origin_from_slice, path_from_file_url,
 };
 
+/// `s3://bucket/key` holds raw key bytes, not a WHATWG URL: `fetch` and
+/// `Request` must not serialize it, or the S3 signer encodes the key twice.
+/// Case-insensitive like [`URL::is_s3`].
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/url/lib.rs]: Cut to one line in 966da26.

**Comment by robobun** [comment]
Status: the diff is ready for review.

Reproduced with the new cases in test/js/bun/s3/s3.test.ts ("s3:// url key encoding"): on the released bun, fetch("s3://bkt/my file.txt") sends GET /bkt/my%2520file.txt while Bun.file sends GET /bkt/my%20file.txt. With this branch all four entry points send the same path.

CI on 966da26 (build 111003): the s3 and fetch suites pass on every lane. The red lanes are unrelated to this change and are reported separately: test-crypto-dh-leak.js (x64-asan RSS bound, also red on the previous build), bun-install-lifecycle-scripts.test.ts (windows aarch64 timeout), and regression/issue/10887.test.ts (batch worker crash on alpine). The binary-size job shows the same +12 MB android and freebsd deltas on unrelated builds against the same canary baseline.
